### PR TITLE
i#5843 scheduler: Add unread_last_record()

### PR DESCRIPTION
**Pull request by derekbruening**
Adds a new scheduler output stream interface unread_last_record(). This queues the just-read record for reading later.  This is meant to aid simulators reading one record ahead to the next instruction, but not yet processing it.  The new interface is limited in that it can only be called once at a time and while the output ordinals are fixed up other metrics like the output last timestamp and the input ordinals are not.

Adds a new record type TRACE_TYPE_INVALID for use as a sentinel.

Adds some sanity tests.

Issue: #5843

### CODE REVIEW

**Comment by derekbruening** [comment]
Hit a presumed-flaky failure in x86-64 off.invar test I don't think we've filed before but it seems familiar: filed as #6303; re-running.

File: clients/drcachesim/scheduler/scheduler.h [hunk 1]
```diff
@@ -556,7 +568,9 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
          * before knowing whether a simulator is on the wrong path or not, this routine
          * supports putting back the current record so that it will be re-provided as
          * the first record after (the outermost) stop_speculation(), if
-         * "queue_current_record" is true.
+         * "queue_current_record" is true.  An alternative way to accomplish this
```
Comment (brettcoon) [inline clients/drcachesim/scheduler/scheduler.h:571]: My reading of the comment suggests that start_speculation w/ queue_current_record=true supports arbitrary nesting, while unread_last_record() does not.  If so, that seems like a strong argument against using unread_last_record() as an alternative here.  Unless start_speculation() will _also_ nest multiple unread last records.
For that matter, it's a little unclear what the interaction is between unread_last_record() and start_speculation().  What happens if you call unread_last_record(), start_speculation(), and then next_record()?  Do you get the unread record or the side stream?

File: clients/drcachesim/scheduler/scheduler.h [hunk 2]
```diff
@@ -1033,6 +1048,10 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
     next_record(output_ordinal_t output, RecordType &record, input_info_t *&input,
                 uint64_t cur_time = 0);
 
+    // Undoes the last read.  May only be called once between next_record() calls.
```
Comment (brettcoon) [inline clients/drcachesim/scheduler/scheduler.h:1051]: As mentioned above, please clarify how this interacts with speculation, including speculation that uses the queue_next_record flag.

**Review by brettcoon — APPROVED** [review]
Mostly clarification requests.

File: clients/drcachesim/scheduler/scheduler.h (hunk 2 above)
Comment (brettcoon) [reply clients/drcachesim/scheduler/scheduler.h:1051]: Probably worth adding a test for this interaction as well.

File: clients/drcachesim/scheduler/scheduler.h [hunk 3]
```diff
@@ -548,6 +548,18 @@ template <typename RecordType, typename ReaderType> class scheduler_tmpl_t {
         virtual stream_status_t
         next_record(RecordType &record, uint64_t cur_time);
 
+        /**
+         * Queues the last-read record returned by next_record() such that it will be
+         * returned on the subsequent call to next_record() when this same input is
+         * active.  May not be called multiple times in a row without an intervening
+         * next_record() call.  This will cause ordinal queries on the current input to be
```
[the remark this answers is not in the document]
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:555]: Done, and done.

File: clients/drcachesim/scheduler/scheduler.h (hunk 1 above)
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:571]: On further examination, you are right: the interactions do get messy.  I think the best thing is to disallowed mixing them: have unread_last_record() fail during speculation and have start_speculation(queue_current_record=true) fail if unread_last_record() was just called.  I documented this and added sanity checks for each failure case.

File: clients/drcachesim/scheduler/scheduler.h (hunk 2 above)
Comment (derekbruening) [reply clients/drcachesim/scheduler/scheduler.h:1051]: Done (disallowed to mix), and done on adding a test of each failure case.

**Pull request merged**